File: tools/benchmark_runtime.py
"""Offline runtime benchmark for the Node and Python implementations.

The Android profile is intentionally a workload profile, not an Android emulator.
It keeps the same operations and data shape while using a smaller Termux-like
workload; real device numbers must be collected on the target device.
"""

from __future__ import annotations

import argparse
import json
import os
from pathlib import Path
import shutil
import statistics
import tempfile
import time

SOURCE_ROOT = Path(__file__).resolve().parents[1] / "py"
import sys
sys.path.insert(0, str(SOURCE_ROOT))

from r2sync.crypto import RcloneBase64
from r2sync.local import _scan_vault
from r2sync.planner import classify_sync_path
from r2sync.remote import _collect_remote_objects


def timed(function):
    started = time.perf_counter()
    value = function()
    return value, (time.perf_counter() - started) * 1000


def median(samples):
    return statistics.median(samples)


def make_dataset(root: Path, file_count: int, content_bytes: int) -> None:
    payload = bytes((index * 31 + 17) & 0xff for index in range(content_bytes))
    for index in range(file_count):
        relative = Path(f"dir-{index % 50:02d}") / f"note-{index:05d}.md"
        target = root / relative
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_bytes(payload)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--profile", choices=("windows", "android-pseudo"), default="windows")
    parser.add_argument("--files", type=int)
    parser.add_argument("--bytes", type=int)
    parser.add_argument("--repeats", type=int, default=5)
    options = parser.parse_args()
    defaults = (1000, 8192) if options.profile == "android-pseudo" else (2000, 16384)
    file_count = options.files or defaults[0]
    content_bytes = options.bytes or defaults[1]
    root = Path(tempfile.mkdtemp(prefix="r2-sync-bench-python-"))

    try:
        if root.exists():
            shutil.rmtree(root)
        root.mkdir(parents=True)
        make_dataset(root, file_count, content_bytes)

        cipher = RcloneBase64("benchmark-password")
        scan_samples = []
        local = None
        for _ in range(options.repeats):
            local, elapsed = timed(lambda: _scan_vault(root))
            scan_samples.append(elapsed)

        paths = list(local)
        remote_items = []
        for relative in paths:
            file_index = int(relative.rsplit("note-", 1)[1][:-3])
            remote_items.append({"key": cipher.encrypt_path(relative), "etag": f"etag-{file_index}"})
        previous = {
            relative: {
                "localMtimeMs": info["mtimeMs"],
                "localSize": info["size"],
                "remoteETag": f"etag-{int(relative.rsplit('note-', 1)[1][:-3])}",
                "localContentHash": "benchmark-hash",
            }
            for relative, info in local.items()
        }

        def decode_and_plan():
            remotes, ignored, conflicts = _collect_remote_objects(
                remote_items, cipher, "", [], [], "error", {"ignored": 0}
            )
            actions = [
                classify_sync_path(relative, local.get(relative), remotes.get(relative), previous.get(relative))
                for relative in sorted(set(local) | set(remotes) | set(previous))
            ]
            return actions, ignored, conflicts

        plan_samples = []
        planned = None
        for _ in range(options.repeats):
            planned, elapsed = timed(decode_and_plan)
            plan_samples.append(elapsed)

        content = bytes(content_bytes)
        crypto_samples = []
        checksum = 0
        crypto_items = min(file_count, 256)
        for _ in range(options.repeats):
            def encrypt_decrypt():
                total = 0
                for _ in range(crypto_items):
                    encrypted = cipher.encrypt_content(content)
                    clear = cipher.decrypt_content(encrypted)
                    total += len(clear) + len(encrypted)
                return total
            value, elapsed = timed(encrypt_decrypt)
            crypto_samples.append(elapsed)
            checksum += value

        print(json.dumps({
            "runtime": "python",
            "profile": options.profile,
            "python": f"{os.sys.version_info.major}.{os.sys.version_info.minor}.{os.sys.version_info.micro}",
            "platform": f"{os.name}-{os.uname().machine if hasattr(os, 'uname') else 'windows'}",
            "dataset": {"files": file_count, "bytesPerFile": content_bytes, "repeats": options.repeats},
            "benchmarks": {
                "scanLocalFiles": {"medianMs": median(scan_samples), "samplesMs": scan_samples, "items": file_count},
                "decodeAndPlanNoop": {"medianMs": median(plan_samples), "samplesMs": plan_samples, "items": file_count, "noop": sum(item["type"] == "NOOP" for item in planned[0])},
                "encryptDecryptContent": {"medianMs": median(crypto_samples), "samplesMs": crypto_samples, "items": crypto_items, "bytesPerSample": content_bytes * crypto_items, "checksum": checksum},
            },
        }, ensure_ascii=False))
    finally:
        shutil.rmtree(root, ignore_errors=True)


if __name__ == "__main__":
    main()

File: tools/benchmark-runtime.mjs
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { performance } from "node:perf_hooks";

import { createCipher } from "../src/crypto.js";
import { createIgnoreMatcher } from "../src/ignore.js";
import { listLocalFiles } from "../src/localFiles.js";
import { planSync } from "../src/sync.js";

const args = new Map();
for (let index = 2; index < process.argv.length; index += 1) {
  if (process.argv[index].startsWith("--")) args.set(process.argv[index], process.argv[index + 1]);
}

const profile = args.get("--profile") ?? "windows";
const defaults = profile === "android-pseudo"
  ? { files: 1000, bytes: 8192, repeats: 5 }
  : { files: 2000, bytes: 16384, repeats: 5 };
const fileCount = Number(args.get("--files") ?? defaults.files);
const contentBytes = Number(args.get("--bytes") ?? defaults.bytes);
const repeats = Number(args.get("--repeats") ?? defaults.repeats);
const root = await fs.mkdtemp(path.join(os.tmpdir(), "r2-sync-bench-node-"));

function median(values) {
  const sorted = [...values].sort((left, right) => left - right);
  return sorted[Math.floor(sorted.length / 2)];
}

async function timed(fn) {
  const started = performance.now();
  const value = await fn();
  return { value, ms: performance.now() - started };
}

async function makeDataset() {
  const payload = Buffer.alloc(contentBytes);
  for (let index = 0; index < payload.length; index += 1) payload[index] = (index * 31 + 17) & 0xff;
  for (let index = 0; index < fileCount; index += 1) {
    const rel = `dir-${String(index % 50).padStart(2, "0")}/note-${String(index).padStart(5, "0")}.md`;
    const target = path.join(root, ...rel.split("/"));
    await fs.mkdir(path.dirname(target), { recursive: true });
    await fs.writeFile(target, payload);
  }
}

async function main() {
  await fs.rm(root, { recursive: true, force: true });
  await fs.mkdir(root, { recursive: true });
  await makeDataset();

  const cipher = await createCipher("benchmark-password");
  const matcher = createIgnoreMatcher([], { vaultPath: root });
  const scanSamples = [];
  let localFiles;
  for (let index = 0; index < repeats; index += 1) {
    const sample = await timed(() => listLocalFiles(root, matcher));
    scanSamples.push(sample.ms);
    localFiles = sample.value;
  }

  const paths = [...localFiles.keys()];
  const encryptedPaths = [];
  for (const relPath of paths) encryptedPaths.push(await cipher.encryptPath(relPath));
  const remoteItems = encryptedPaths.map((key, index) => {
    const fileIndex = Number(paths[index].match(/note-(\d+)\.md$/)[1]);
    return { key, etag: `etag-${fileIndex}` };
  });
  const prevEntries = {};
  for (const [relPath, info] of localFiles) {
    const index = Number(relPath.match(/note-(\d+)\.md$/)[1]);
    prevEntries[relPath] = {
      localMtimeMs: info.mtimeMs,
      localSize: info.size,
      remoteETag: `etag-${index}`,
      localContentHash: "benchmark-hash",
    };
  }

  const planSamples = [];
  let planned;
  for (let index = 0; index < repeats; index += 1) {
    const sample = await timed(() => planSync({
      vaultPath: root,
      localFiles,
      remoteItems,
      prevEntries,
      ignoreMatcher: matcher,
      cipher,
      r2Prefix: "",
      r2: { getObject: async () => { throw new Error("not expected in noop benchmark"); } },
    }));
    planSamples.push(sample.ms);
    planned = sample.value;
  }

  const content = Buffer.alloc(contentBytes);
  const cryptoSamples = [];
  let checksum = 0;
  for (let index = 0; index < repeats; index += 1) {
    const sample = await timed(async () => {
      let total = 0;
      for (let item = 0; item < Math.min(fileCount, 256); item += 1) {
        const encrypted = await cipher.encryptContent(content);
        const clear = await cipher.decryptContent(encrypted);
        total += clear.length + encrypted.length;
      }
      return total;
    });
    cryptoSamples.push(sample.ms);
    checksum += sample.value;
  }

  console.log(JSON.stringify({
    runtime: "node",
    profile,
    node: process.version,
    platform: `${process.platform}-${process.arch}`,
    dataset: { files: fileCount, bytesPerFile: contentBytes, repeats },
    benchmarks: {
      scanLocalFiles: { medianMs: median(scanSamples), samplesMs: scanSamples, items: fileCount },
      decodeAndPlanNoop: { medianMs: median(planSamples), samplesMs: planSamples, items: fileCount, noop: planned.filter((item) => item.type === "NOOP").length },
      encryptDecryptContent: { medianMs: median(cryptoSamples), samplesMs: cryptoSamples, items: Math.min(fileCount, 256), bytesPerSample: contentBytes * Math.min(fileCount, 256), checksum },
    },
  }));
}

try {
  await main();
} finally {
  await fs.rm(root, { recursive: true, force: true });
}
